# **WEEK 5 Assignment – Introduction to Deep Learning & Neural Networks with Keras**
Surya Prakash Karasala
 
Date: 03/22/2021

## **Importing Required Packages**

Let's start by importing the <em>pandas</em> and the Numpy libraries.

In [1]:
import pandas as pd
import numpy as np

We will be using the dataset provided in the assignment

<strong>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:</strong>

<strong>1. Cement</strong>

<strong>2. Blast Furnace Slag</strong>

<strong>3. Fly Ash</strong>

<strong>4. Water</strong>

<strong>5. Superplasticizer</strong>

<strong>6. Coarse Aggregate</strong>

<strong>7. Fine Aggregate</strong>

## Loading Data

Let's read the dataset into a pandas dataframe.

In [2]:
concrete_data = pd.read_csv('https://cocl.us/concrete_data')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  ...  Fine Aggregate  Age  Strength
0   540.0                 0.0      0.0  ...           676.0   28     79.99
1   540.0                 0.0      0.0  ...           676.0   28     61.89
2   332.5               142.5      0.0  ...           594.0  270     40.27
3   332.5               142.5      0.0  ...           594.0  365     41.05
4   198.6               132.4      0.0  ...           825.5  360     44.30

[5 rows x 9 columns]

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa. 

###**Let's check how many data points we have**.

In [ ]:
concrete_data.shape

(1030, 9)

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.

Let's check the dataset for any missing values.

In [3]:
concrete_data.describe()

Cement  Blast Furnace Slag  ...          Age     Strength
count  1030.000000         1030.000000  ...  1030.000000  1030.000000
mean    281.167864           73.895825  ...    45.662136    35.817961
std     104.506364           86.279342  ...    63.169912    16.705742
min     102.000000            0.000000  ...     1.000000     2.330000
25%     192.375000            0.000000  ...     7.000000    23.710000
50%     272.900000           22.000000  ...    28.000000    34.445000
75%     350.000000          142.950000  ...    56.000000    46.135000
max     540.000000          359.400000  ...   365.000000    82.600000

[8 rows x 9 columns]

In [4]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.

### **Cleaning and Normalizing Data**

The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.

In [5]:
concrete_data_columns = concrete_data.columns
predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

Let's do a quick sanity check of the predictors and the target dataframes.

In [6]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  ...  Coarse Aggregate  Fine Aggregate  Age
0   540.0                 0.0      0.0  ...            1040.0           676.0   28
1   540.0                 0.0      0.0  ...            1055.0           676.0   28
2   332.5               142.5      0.0  ...             932.0           594.0  270
3   332.5               142.5      0.0  ...             932.0           594.0  365
4   198.6               132.4      0.0  ...             978.4           825.5  360

[5 rows x 8 columns]

In [7]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

In [8]:
n_cols = predictors.shape[1] # number of predictors
n_cols

8

<a id="item1"></a>

<a id="item1"></a>

## Import Keras

#### Let's go ahead and import the Keras library

In [9]:
import keras

As you can see, the TensorFlow backend was used to install the Keras library.

Let's import the rest of the packages from the Keras library that we will need to build our regressoin model.

In [10]:
from keras.models import Sequential
from keras.layers import Dense

In [11]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

The above function creates a model that has one hidden layer with 10 neurons and a ReLU activation function. It uses the adam optimizer and the mean squared error as the loss function.

Let's import scikit-learn in order to randomly split the data into a training and test sets

In [12]:
from sklearn.model_selection import train_test_split

Splitting the data into a training and test sets by holding 30% of the data for testing

In [13]:
X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=42)

## Train and Test the Network

Let's call the function now to create our model.

In [14]:
# build the model
model = regression_model()

Next, we will train the model for 50 epochs.


In [15]:
# fit the model
epochs = 50
model.fit(X_train, y_train, epochs=epochs, verbose=1)

Epoch 1/50
23/23 [==============================] - 1s 1ms/step - loss: 56021.4193
Epoch 2/50
23/23 [==============================] - 0s 1ms/step - loss: 11305.9522
Epoch 3/50
23/23 [==============================] - 0s 1ms/step - loss: 3379.8955
Epoch 4/50
23/23 [==============================] - 0s 1ms/step - loss: 3294.6796
Epoch 5/50
23/23 [==============================] - 0s 1ms/step - loss: 2993.1638
Epoch 6/50
23/23 [==============================] - 0s 1ms/step - loss: 2964.6388
Epoch 7/50
23/23 [==============================] - 0s 1ms/step - loss: 2841.1146
Epoch 8/50
23/23 [==============================] - 0s 975us/step - loss: 2638.5429
Epoch 9/50
23/23 [==============================] - 0s 1ms/step - loss: 2303.5320
Epoch 10/50
23/23 [==============================] - 0s 1ms/step - loss: 2512.1062
Epoch 11/50
23/23 [==============================] - 0s 1ms/step - loss: 2191.4062
Epoch 12/50
23/23 [==============================] - 0s 1ms/step - loss: 2166.2650
Epoch 13/

Next we need to evaluate the model on the test data.

In [16]:
loss_val = model.evaluate(X_test, y_test)
y_pred = model.predict(X_test)
loss_val

10/10 [==============================] - 0s 1ms/step - loss: 312.7010


312.7009582519531

Now we need to compute the mean squared error between the predicted concrete strength and the actual concrete strength.

Let's import the mean_squared_error function from Scikit-learn.

In [17]:
from sklearn.metrics import mean_squared_error

In [18]:
mean_square_error = mean_squared_error(y_test, y_pred)
mean = np.mean(mean_square_error)
standard_deviation = np.std(mean_square_error)
print(mean, standard_deviation)

312.7009032587031 0.0


Create a list of 50 mean squared errors and report mean and the standard deviation of the mean squared errors.

In [19]:
total_mean_squared_errors = 50
epochs = 50
mean_squared_errors = []
for i in range(0, total_mean_squared_errors):
    X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    y_pred = model.predict(X_test)
    mean_square_error = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square_error)

mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print('\n')
print("Below is the mean and standard deviation of " +str(total_mean_squared_errors) + " mean squared errors without normalized data. Total number of epochs for each training is: " +str(epochs) + "\n")
print("Mean: "+str(mean))
print("Standard Deviation: "+str(standard_deviation))

MSE 1: 121.95288848876953
MSE 2: 145.2391357421875
MSE 3: 114.91912078857422
MSE 4: 130.37527465820312
MSE 5: 125.41300964355469
MSE 6: 109.977294921875
MSE 7: 142.4945068359375
MSE 8: 108.30347442626953
MSE 9: 122.76383972167969
MSE 10: 118.01298522949219
MSE 11: 102.64520263671875
MSE 12: 126.24449920654297
MSE 13: 121.47888946533203
MSE 14: 116.92609405517578
MSE 15: 109.37126922607422
MSE 16: 109.85301971435547
MSE 17: 105.96543884277344
MSE 18: 95.43729400634766
MSE 19: 95.0811538696289
MSE 20: 114.39599609375
MSE 21: 96.89361572265625
MSE 22: 105.53860473632812
MSE 23: 112.22298431396484
MSE 24: 106.6091079711914
MSE 25: 108.7795181274414
MSE 26: 103.71099853515625
MSE 27: 128.18467712402344
MSE 28: 109.34622955322266
MSE 29: 111.38359832763672
MSE 30: 112.36216735839844
MSE 31: 135.80026245117188
MSE 32: 108.52020263671875
MSE 33: 102.13824462890625
MSE 34: 115.26837921142578
MSE 35: 114.88044738769531
MSE 36: 134.95152282714844
MSE 37: 121.79768371582031
MSE 38: 125.59642028808